src/Interpreters/TemporaryDataOnDisk.cpp

```diff
@@ -385,4 +387,105 @@ TemporaryFileStream::~TemporaryFileStream()
     }
 }
 
+ThreadSafeTemporaryFileStream::ThreadSafeTemporaryFileStream(const Block & header_, WriteBufferPtr file_buf_, size_t max_block_size_)
+    : header(header_)
+    , file_out_buf(file_buf_)
+    , file_in_buf(nullptr)
+    , max_block_size(max_block_size_)
+{
+}
+
+void ThreadSafeTemporaryFileStream::finishWriting()
+{
+    flushPendingBlocks();
+    std::lock_guard lock(file_mutex);
+    if (write_finished)
+    {
+        return;
+    }
+    write_finished = true;
+    file_out_buf->finalize();
+}
+
+bool ThreadSafeTemporaryFileStream::isWriteFinished() const
+{
+    return write_finished;
+}
+
+bool ThreadSafeTemporaryFileStream::isEof() const
+{
+    std::lock_guard lock(file_mutex);
+    return write_finished && file_in_buf && file_in_buf->eof();
+}
+
+void ThreadSafeTemporaryFileStream::write(const Block & block)
+{
+    if (write_finished) [[unlikely]]
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Writing a finished file");
+    if (!block.rows())
+        return;
+    bool need_flush = false;
+    {
+        std::lock_guard pending_blocks_lock(pending_blocks_mutex);
+        pending_rows += block.rows();
+        pending_blocks.push_back(block);
```

Comment:
the block need to be moved, not copied. Blocks we gonna spill atm are not referenced anywhere else.

Reply:
The `block` is a const reference, could not be moved

Reply:
After scattering the current block by hash, one bucket becoming current and the rest are spilled to disk. If you check the code path for blocks which should be spilled, you'll see that they are passed by reference and only copied here into pending blocks. So, I'm suggesting avoiding unnecessary copy here. It wasn't the case before, since there were no pending blocks.

Reply:
Have updated related interfaces to accept right reference